[TT-549] Document how to write a spot linter (#552)
airtasker/spot: lib/src/linting/README.MD

# Linting

> Spot linters are a set of functions whose goal is to ensure a contract definition conforms with our API guidelines. In other words, linters check whether or not a contract is semantically correct.

Reviewer: ```suggestion
Spot linters are a set of functions whose goal is to ensure a contract definition conforms with our API guidelines.
```
I'm not sure we should talk about `semantically correct`. Linters could just describe an API guideline that is just representing a standard at Airtasker. Which at any point in time could be a workaround for a certain situation

Linting rules can be found in files inside the spot directory (under [lib/src/linting](/lib/src/linting). Each file corresponds to a specific rule and holds the logic to validate a subset of a contract (e.g. headers definition, request/response types, etc).

## Anatomy of a linting rule

> Let's take an example of a linting rule to demonstrate how we can catch semantic errors in a given contract  :

Reviewer: ```suggestion
Let's take an example of a linting rule to demonstrate how we can catch errors in a given contract:
```

import { noNullableFieldsWithinRequests } from "./rules/no-nullable-fields-within-requests";
import { noObjectsInQueryParams } from "./rules/no-objects-in-query-params";
import { noOmittableFieldsWithinResponses } from "./rules/no-omittable-fields-within-responses";
import { oneSuccessResponsePerEndpoint } from "./rules/one-success-response-per-endpoint";
import { yourLintingRuleName } from "./your-linting-rule-name";

export const availableRules = {
  "has-discriminator": hasDiscriminator,
  "has-request-payload": hasRequestPayload,
  "has-response-payload": hasResponsePayload,
  "no-nested-types-within-unions": noNestedTypesWithinUnions,
  "no-nullable-arrays": noNullableArrays,
  "no-nullable-fields-within-requests": noNullableFieldsWithinRequests,
  "no-objects-in-query-params": noObjectsInQueryParams,
  "no-omittable-fields-within-responses": noOmittableFieldsWithinResponses,
  "one-success-response-per-endpoint": oneSuccessResponsePerEndpoint
  "yourLintingRuleName": yourLintingRuleName
};

export type RuleName = keyof typeof availableRules;
```

To run the linter against a contract, simply do

```
$ ./bin/run lint [contract].ts
```

## Test a linting rule

> Testing a linting rule is necessary. It allows to ensure it works as expected and comply with the guidelines.

Reviewer: ```suggestion
Testing a linting rule is necessary. It makes sure everything works as expected and complies with the guidelines.
```

> Every linting rule comes with a specification file. Writing tests for a specific rule consists in creating contracts inputs and programmatically call the function associated with this rule.

Reviewer: ```suggestion
Every linting rule comes with a specification file. Writing tests for a specific rule consists in creating contracts inputs and programmatically calling the function associated with this rule.
```
